### PR TITLE
Animated List Detail Composable

### PR DESCRIPTION
**Pull request by BottleRocket-Colin**
Animated List Detail

This PR covers the creation of the AnimatedListDetail composable and the implementation of it within the Arch Demo.  Once final, we can pull the AnimatedListDetail code into Launchpad library to make it more available.

Note:  Since the Arch Demo did not orginally have a "view" snippet functionality I've left that part undone.  That could be a good work assignment for devs coming onto the bench as an intro to Compose.

### CODE REVIEW

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetListDetailComposable.kt [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+                    // Show Create snippet in detail pane when applicable
+                    if (it == CreateSnippetItem) {
+                        val createSnippetViewModel: CreateSnippetViewModel = getViewModel()
+                        CreateSnippetScreen(state = createSnippetViewModel.toState())
+                        createSnippetViewModel.onSuccess.LaunchCollection {
+                            select(null)
+                            snippetsViewModel.refreshSnippets()
+                        }
+                    } else {
+                        //    TODO - Normal Detail screen.
```
Comment (BottleRocket-Colin) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetListDetailComposable.kt:85]: This is where we can add a "detail" screen for snippets that gives a little more info on files, etc.  This will require an additional call to be made via BitBucket Repo.

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/util/NavUtils.kt [hunk 2]
```diff
@@ -15,7 +11,3 @@ fun NavOptionsBuilder.popToMainInclusive() {
         inclusive = true
     }
 }
-
```
Comment (BottleRocket-Colin) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/util/NavUtils.kt:18]: Moved to compose module

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/ComposeActivity.kt [hunk 3]
```diff
@@ -122,10 +122,12 @@ class ComposeActivity : ComponentActivity() {
             NavItemState(
                 icon = R.drawable.ic_snippet,
                 itemText = R.string.snippets_title,
-                selected = currentRoute == Routes.Snippets
+                // selected = currentRoute == Routes.Snippets
```
Comment (BottleRocket-Colin) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/ComposeActivity.kt]: Left old routes temporarily, for debugging/regression testing purposes.

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/MainNavGraph.kt [hunk 4]
```diff
@@ -260,6 +261,7 @@ fun NavGraphBuilder.mainNavGraph(navController: NavController, activity: Compose
             snippetsComposable(this, navController)
             createSnippetComposable(this, navController)
             profileComposable(this, navController)
+            newSnippetsComposable(this)
```
Comment (BottleRocket-Colin) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/MainNavGraph.kt]: Should be able to rename this after we remove old code.  Perhaps a new PR after everything is final?

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/Routes.kt [hunk 5]
```diff
@@ -13,6 +13,7 @@ object Routes {
     const val Snippets = "snippets"
     const val CreateSnippet = "snippets/create"
     const val Profile = "profile"
+    const val NewSnippet = "newsnippets"
```
Comment (BottleRocket-Colin) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/Routes.kt]: Will cleanup old snippets code and rename in new PR after code is final.

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/SnippetUiModel.kt [hunk 6]
```diff
@@ -3,6 +3,8 @@ package com.bottlerocketstudios.compose.snippets
 import com.bottlerocketstudios.compose.util.StringIdHelper
 
 data class SnippetUiModel(
+    val id: String,
+    val workspaceSlug: String,
     val title: String,
```
Comment (BottleRocket-Colin) [inline compose/src/main/java/com/bottlerocketstudios/compose/snippets/SnippetUiModel.kt:8]: Added these in preparation for Snippet Detail view, but it got complictaed enough that I felt it should be a different PR.

File: data/src/main/java/com/bottlerocketstudios/brarchitecture/data/model/SnippetDto.kt [hunk 7]
```diff
@@ -9,9 +9,12 @@ import java.time.ZonedDateTime
 @JsonClass(generateAdapter = true)
 @Parcelize
 data class SnippetDto(
-    @Json(name = "id") val id: String? = null,
-    @Json(name = "title") val title: String? = null,
-    @Json(name = "is_private") val isPrivate: Boolean? = null,
-    @Json(name = "owner") val owner: UserDto? = null,
-    @Json(name = "updated_on") val updated: ZonedDateTime? = null,
+    val id: String? = null,
```
Comment (BottleRocket-Colin) [inline data/src/main/java/com/bottlerocketstudios/brarchitecture/data/model/SnippetDto.kt]: Question:  Were the Json annotations auto generated and/or code standard?  I usually only use those when JSON name doesn't follow Kotlin variable naming standards, there is a mispelled word, etc.  I have no issue with adding them back in for all variables, just curious.

File: .idea/codeStyles/Project.xml [hunk 8]
```diff
@@ -108,7 +108,6 @@
                   <XML_NAMESPACE>http://schemas.android.com/apk/res/android</XML_NAMESPACE>
                 </AND>
               </match>
-              <order>ANDROID_ATTRIBUTE_ORDER</order>
```
Comment (jaredstockton) [inline .idea/codeStyles/Project.xml:111]: @BottleRocket-Colin Was this intentional? I'd prefer to keep it as is unless it was removed by some newer version of Android Studio

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/MainNavGraph.kt (hunk 4 above)
Comment (jaredstockton) [reply app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/MainNavGraph.kt]: If this is going straight into `compose/develop`, I'd be in favor of refactoring in this PR. If we create a long running branch (similar to the old `compose/navigation` branch), then it could be in a future PR.

Regardless, are you thinking of something along the lines of `snippetListDetailComposable` or something else?

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetListDetailComposable.kt [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+        val config = LocalConfiguration.current
+
+        val list = snippetsViewModel.snippets.collectAsState(initial = emptyList())
+        controls.title = stringResource(id = R.string.snippets_title)
+        controls.topLevel = true
+
+        AnimatedListDetail(
+            list = list.value + CreateSnippetItem ,
+            keyProvider = { it.id },
+            smallScreen = config.smallestScreenWidthDp < 580) {
```
Comment (jaredstockton) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetListDetailComposable.kt]: @BottleRocket-Colin It would be nice if we followed the terminology of `Compact`, `Medium`, and `Expanded` as mentioned on https://developer.android.com/guide/topics/large-screens/support-different-screen-sizes

1. How about using `compact` instead of `small`?
2. How do you feel about using `window` instead of `screen` values? Meaning that if an app is being displayed in multi window mode, the actual screen values won't matter as much as the window values
3. How do you feel about using 600 (breakpoint value for `Medium`) instead of 580?

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetsViewModel.kt [hunk 10]
```diff
@@ -19,13 +20,17 @@ class SnippetsViewModel : BaseViewModel() {
...
+                id = it.id ?: "",
+                workspaceSlug = it.workspace?.slug ?: "",
                 title = it.title ?: "",
                 userName = it.owner?.displayName ?: "",
                 formattedLastUpdatedTime = it.updated.formattedUpdateTime(clock = clock)
             )
         }
     }
 
+    val showCreateCTA = MutableStateFlow(true)
```
Comment (jaredstockton) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetsViewModel.kt]: @BottleRocket-Colin nit: Refactor to `showCreateCta` (per guidance at https://developer.android.com/kotlin/style-guide#camel_case)

File: buildSrc/src/main/kotlin/Dependencies.kt [hunk 11]
```diff
@@ -171,7 +171,8 @@ private object Libraries {
 
     // Accompanist
     // https://github.com/google/accompanist/releases
-    const val ACCOMPANIST = "com.google.accompanist:accompanist-webview:0.24.5-alpha"
+    const val ACCOMPANIST_WEBVIEW = "com.google.accompanist:accompanist-webview:0.24.5-alpha"
```
Comment (jaredstockton) [inline buildSrc/src/main/kotlin/Dependencies.kt]: As there are now multiple dependencies relying on the same version number, go ahead and create a const val holding the version and replace all hardcoded version numbers with that property reference.

It's nice to have a single source of truth to update when new dependency versions are available.

```
private const val ACCOMPANIST_VERSION = "0.24.5-alpha"
```

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetListDetailComposable.kt [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+                                select("CREATE_SNIPPET_SCREEN")
+                            }
+                        }
+                    )
+                )
+            }
+
+            // Define Detail UI for create, detail, and empty states
+            Detail { model ->
+                model.also {
```
Comment (jaredstockton) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetListDetailComposable.kt]: @BottleRocket-Colin Should you be using `model?.also {` here since you handle null in the trailing `run` block? This would prevent null from being handled in both the `also` block and the trailing `run` block.

File: data/src/main/java/com/bottlerocketstudios/brarchitecture/data/model/SnippetDto.kt (hunk 7 above)
Comment (jaredstockton) [reply data/src/main/java/com/bottlerocketstudios/brarchitecture/data/model/SnippetDto.kt]: @BottleRocket-Colin I don't remember how the data classes were originally built but they didn't define the json name explicitly until https://github.com/BottleRocketStudios/Android-ArchitectureDemo/commit/ce77220f6cc84add85e7e52662628bff18e6d9c9

> Explicitly specify @Json(name = ... ) for all network request/response data classes
> 
> Doing so provides the following benefits:
> 
> 1. Supports correct (property) naming conventions in the app regardless of serialization key naming conventions.
> 2. Prevents future property name refactoring breaking serialization

I'd say add them back to prevent against future runtime failures when/if the Dto properties are renamed.

File: gradle.properties [hunk 13]
```diff
@@ -22,7 +22,7 @@ org.gradle.configureondemand=false
 kotlin.code.style=official
 # Using AndroidX: https://developer.android.com/jetpack/androidx#using_androidx
 android.useAndroidX=true
-android.enableJetifier=true
+android.enableJetifier=false
```
Comment (jaredstockton) [inline gradle.properties:23]: Woohoo!

**Review by jaredstockton — COMMENTED** [review]
@BottleRocket-Colin Left some comments to address. Overall, it's looking pretty good!

File: .idea/codeStyles/Project.xml (hunk 8 above)
Comment (BottleRocket-Colin) [reply .idea/codeStyles/Project.xml:111]: Not intentional could have happened from switch different versions of Studio??

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetListDetailComposable.kt (hunk 9 above)
Comment (BottleRocket-Colin) [reply app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetListDetailComposable.kt]: I'll look into window size class usage.

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetsViewModel.kt (hunk 10 above)
Comment (BottleRocket-Colin) [reply app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetsViewModel.kt]: Will do

File: buildSrc/src/main/kotlin/Dependencies.kt (hunk 11 above)
Comment (BottleRocket-Colin) [reply buildSrc/src/main/kotlin/Dependencies.kt]: will do

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetListDetailComposable.kt (hunk 12 above)
Comment (BottleRocket-Colin) [reply app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetListDetailComposable.kt]: The also block returns this (which is why this works right now), but you're correct in that we don't want to run the code in the empty else block.  No current bug, but would have been very strange when we added that section in.  Good spot!!

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetListDetailComposable.kt (hunk 9 above)
Comment (jaredstockton) [reply app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetListDetailComposable.kt]: Cool! I see you've added the enum to represent the different types. Last thing I see for this PR is to add logic to determine the current window type and then use it instead of the isSmallScreen boolean

**Comment by BottleRocket-Colin** [comment]
@br-jared-stockton - Should have everything for review again.

**Comment by jaredstockton** [comment]
> @br-jared-stockton - Should have everything for review again.

It all looks good! This is great to have support in the arch demo now to handle multiple screen sizes :D

**Review by jaredstockton — APPROVED** [review]
